`Internal error: during type inference` on v1.12

On a private package (i will try to create a shareable reproduction or get a bisect), with Julia v1.12 (7 day old master, at `[1b0fba7](https://github.com/JuliaLang/julia/commit/1b0fba7ffd455d919f1c4f6bb8f31d7c4fae0cc0)`)  i see the following:
```julia
 Internal error: during type inference of
 (::TaskGroups.var"#_maybe_spawn_worker##0#_maybe_spawn_worker##1"{TaskGroups.TaskGroup})()
 Encountered unexpected error in runtime:
 MethodError(f=Base.string, args=(Expr(:call, :<=, :id, Expr(:call, :length, Expr(:., :compact, :(:pending_nodes)))),), world=0x0000000000001961)
 thread (4) jl_method_error_bare at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:2507
 thread (4) jl_method_error at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:2525
 thread (4) jl_f_throw_methoderror at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/builtins.c:587 [inlined]
 thread (4) jl_f_throw_methoderror at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/builtins.c:583
 thread (4) _assert_tostring at ./error.jl:249
 thread (4) unknown function (ip: 0x7fffbce66dd2) at (unknown file)
 thread (4) did_already_insert at ./../usr/share/julia/Compiler/src/ssair/passes.jl:373
 thread (4) process_phinode_values at ./../usr/share/julia/Compiler/src/ssair/ir.jl:1281
 thread (4) jfptr_process_phinode_values_98514.1 at /nix/store/72g8yzry2c5vfsa161p2j4xbigrwpgbm-julia-1.12.0-DEV/lib/julia/sys.so (unknown line)
 thread (4) process_node! at ./../usr/share/julia/Compiler/src/ssair/ir.jl:1629
 thread (4) iterate_compact at ./../usr/share/julia/Compiler/src/ssair/ir.jl:1937
 thread (4) iterate at ./../usr/share/julia/Compiler/src/ssair/ir.jl:1859 [inlined]
 thread (4) compact! at ./../usr/share/julia/Compiler/src/ssair/ir.jl:2136
 thread (4) run_passes_ipo_safe at ./../usr/share/julia/Compiler/src/optimize.jl:1013
 thread (4) run_passes_ipo_safe at ./../usr/share/julia/Compiler/src/optimize.jl:1027 [inlined]
 thread (4) optimize at ./../usr/share/julia/Compiler/src/optimize.jl:1002
 thread (4) jfptr_optimize_118359.1 at /nix/store/72g8yzry2c5vfsa161p2j4xbigrwpgbm-julia-1.12.0-DEV/lib/julia/sys.so (unknown line)
 thread (4) finish_nocycle at ./../usr/share/julia/Compiler/src/typeinfer.jl:203
 thread (4) jfptr_finish_nocycle_119415.1 at /nix/store/72g8yzry2c5vfsa161p2j4xbigrwpgbm-julia-1.12.0-DEV/lib/julia/sys.so (unknown line)
 thread (4) typeinf at ./../usr/share/julia/Compiler/src/abstractinterpretation.jl:4275
 thread (4) typeinf_ext at ./../usr/share/julia/Compiler/src/typeinfer.jl:1154
 thread (4) typeinf_ext_toplevel at ./../usr/share/julia/Compiler/src/typeinfer.jl:1228
 thread (4) jfptr_typeinf_ext_toplevel_119358.1 at /nix/store/72g8yzry2c5vfsa161p2j4xbigrwpgbm-julia-1.12.0-DEV/lib/julia/sys.so (unknown line)
 thread (4) jl_apply at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/julia.h:2248 [inlined]
 thread (4) jl_type_infer at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:452
 thread (4) jl_compile_method_internal at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:2971
 thread (4) _jl_invoke at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:3465 [inlined]
 thread (4) ijl_apply_generic at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/gf.c:3673
 thread (4) jl_apply at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/julia.h:2248 [inlined]
 thread (4) start_task at /build/julia-v1.12.0-DEV+RAI-d8e121836bcd203fd304b278e19b5a76db4b811f/src/task.c:1277
```
which points at https://github.com/JuliaLang/julia/blob/1b0fba7ffd455d919f1c4f6bb8f31d7c4fae0cc0/Compiler/src/ssair/passes.jl#L373

slightly more reduced by hand
```julia
julia> function _worker_task()                                                                                                  
           while true                                                                                                           
               r = let                                                                                                          
               try                                                                                                              
                   if @noinline rand(Bool)                                                                                      
                       return nothing                                                                                           
                   end                                                                                                          
                   q, m                                                                                                                                                                                                                                          
               finally                                                                                                          
                   missing                                                                                                      
               end                                                                                                              
               end                                                                                                              
               r[1]::Bool                                                                                                       
           end                                                                                                                  
       end; code_typed(_worker_task, ())                                                                                        
```

fails with
```
ERROR: AssertionError: id <= length(compact.pending_nodes)
2 <= 0
```
with IncrementalCompact state
```
2  1 ── %1  = φ (#11 => Compiler.OldSSAValue(17))::Core.Const(nothing)      │
   │    %2  = φ (#11 => Compiler.OldSSAValue(18))::Bool                     │                                                   
   └───       φ (#11 => Compiler.OldSSAValue(21))::Bool                     │                                                   
   2 ──       nothing::Nothing                                              │                                                   
4  3 ── %5  = enter #12                                                     │                                                   
5  4 ── %6  =    invoke Main.rand(Main.Bool::Type{Bool})::Bool              │
   └───       goto #7 if not %6                                             │                                                                                                                                                                                    
6  5 ── %8  = $(Expr(:leave, :(%5)))                                        │
   6 ──       goto #9                                                       │                                                                                                                                                                                    
8  7 ──       Main.q::Any                                                   │
   │          Main.m::Any                                                   │
   └─── %12 = $(Expr(:leave, :(%5)))                                        │                                                                                                                                                                                    
   8 ──       nothing::Nothing                                              │
10 9 ┄─       φ (#6 => 1, #8 => -1)::Int64                                  │                                                                                                                                                                                    
   │          φ (#6 => Main.nothing, #8 => %1)::Core.Const(nothing)         │
   │          φ (#6 => true, #8 => %2)::Bool                                │
─────────────────────────────────────────────────────────────────────────────
   │    %19 = φ (#6 => Compiler.OldSSAValue(32), #8 => %11)::Any            │
   │          nothing::Tuple{Any, Any}                                      │
   │    %21 = φ (#6 => %4, #8 => true)::Bool                                │
   │    %22 =   builtin (%16 === 1)::Bool                                   │
   └───       goto #11 if not %22                                           │                                                                                                                                                                                    
   10 ─       $(Expr(:throw_undef_if_not, Symbol(""), :(%18)))::Any         │
   └───       return %17                                                    │
   11 ─       $(Expr(:throw_undef_if_not, Symbol(""), :(%21)))::Any         │
   │            builtin Core.typeassert(%19, Main.Bool)::Bool               │
   └───       goto #1                                                       │
   12 ┄          invoke Base.rethrow()::Union{}                             │                        
   └───       unreachable                                                   │
```

note that this is perhaps is a bit confusing, since inference thinks there is a loop-carried dependency on the value and nullstate of the return value slot, both of which start initialized to #undef, but dynamically this should be okay

--- a/Compiler/src/ssair/inlining.jl
+++ b/Compiler/src/ssair/inlining.jl
@@ -126,10 +126,11 @@ function cfg_inline_item!(ir::IRCode, idx::Int, todo::InliningTodo, state::CFGIn
     block = block_for_inst(ir, idx)
     inline_into_block!(state, block)
 
-    if !isempty(inlinee_cfg.blocks[1].preds)
+    if length(inlinee_cfg.blocks[1].preds) > 1
         need_split_before = true
+    else
+        @assert inlinee_cfg.blocks[1].preds[1] == 0
     end
-
     last_block_idx = last(state.cfg.blocks[block].stmts)
     if false # TODO: ((idx+1) == last_block_idx && isa(ir[SSAValue(last_block_idx)], GotoNode))
         need_split = false
@@ -166,12 +167,15 @@ function cfg_inline_item!(ir::IRCode, idx::Int, todo::InliningTodo, state::CFGIn
     end
     new_block_range = (length(state.new_cfg_blocks)-length(inlinee_cfg.blocks)+1):length(state.new_cfg_blocks)
 
-    # Fixup the edges of the newely added blocks
+    # Fixup the edges of the newly added blocks
     for (old_block, new_block) in enumerate(bb_rename_range)
         if old_block != 1 || need_split_before
             p = state.new_cfg_blocks[new_block].preds
             let bb_rename_range = bb_rename_range
                 map!(p, p) do old_pred_block
+                    if old_block == 1 && old_pred_block == 0
+                        return first(bb_rename_range) - 1
+                    end
                     return old_pred_block == 0 ? 0 : bb_rename_range[old_pred_block]
                 end
             end
@@ -184,10 +188,6 @@ function cfg_inline_item!(ir::IRCode, idx::Int, todo::InliningTodo, state::CFGIn
                 end
             end
         end
-    end
-
-    if need_split_before
-        push!(state.new_cfg_blocks[first(bb_rename_range)].preds, first(bb_rename_range)-1)
     end
 
     any_edges = false
@@ -399,7 +399,7 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
     else
         bb_offset, post_bb_id = popfirst!(todo_bbs)
         # This implements the need_split_before flag above
-        need_split_before = !isempty(item.ir.cfg.blocks[1].preds)
+        need_split_before = length(item.ir.cfg.blocks[1].preds) > 1
         if need_split_before
             finish_current_bb!(compact, 0)
         end

--- a/Compiler/src/ssair/inlining.jl
+++ b/Compiler/src/ssair/inlining.jl
@@ -173,9 +173,12 @@ function cfg_inline_item!(ir::IRCode, idx::Int, todo::InliningTodo, state::CFGIn
             p = state.new_cfg_blocks[new_block].preds
             let bb_rename_range = bb_rename_range
                 map!(p, p) do old_pred_block
+                    # the meaning of predecessor 0 depends on the block we encounter it:
+                    #   - in the first block, it represents the function entry and so needs to be re-mapped
                     if old_block == 1 && old_pred_block == 0
                         return first(bb_rename_range) - 1
                     end
+                    #   - elsewhere, it represents external control-flow from a caught exception which is un-affected by inlining
                     return old_pred_block == 0 ? 0 : bb_rename_range[old_pred_block]
                 end
             end

--- a/Compiler/src/ssair/ir.jl
+++ b/Compiler/src/ssair/ir.jl
@@ -105,6 +105,9 @@ function compute_basic_blocks(stmts::Vector{Any})
     end
     # Compute successors/predecessors
     for (num, b) in enumerate(blocks)
+        if b.stmts.start == 1
+            push!(b.preds, 0) # the entry block has a virtual predecessor
+        end
         terminator = stmts[last(b.stmts)]
         if isa(terminator, ReturnNode)
             # return never has any successors

--- a/Compiler/test/ssair.jl
+++ b/Compiler/test/ssair.jl
@@ -825,3 +825,23 @@ end
 
 @test_throws ErrorException Base.code_ircode(+, (Float64, Float64); optimize_until = "nonexisting pass name")
 @test_throws ErrorException Base.code_ircode(+, (Float64, Float64); optimize_until = typemax(Int))
+
+#57153 check that the CFG has a #0 block predecessor and that we don't fail to compile code that observes that
+function _worker_task57153()
+    while true
+        r = let
+        try
+            if @noinline rand(Bool)
+                return nothing
+            end
+            q, m
+        finally
+            missing
+        end
+        end
+        r[1]::Bool
+    end
+end
+let ir = Base.code_ircode(_worker_task57153, (), optimize_until="CC: COMPACT_2")[1].first
+    @test findfirst(x->x==0, ir.cfg.blocks[1].preds) !== nothing
+end

